lib/runtime/wasmer/imports.go

```diff
@@ -185,45 +188,191 @@ func ext_sandbox_memory_teardown_version_1(context unsafe.Pointer, a C.int32_t)
+	memory := instanceContext.Memory().Data()
+
+	keys := runtimeCtx.Keystore.Ed25519PublicKeys()
+
+	ptr, err := runtimeCtx.Allocator.Allocate(uint32(len(keys) * 32))
+	if err != nil {
+		logger.Error("[ext_crypto_ed25519_public_keys_version_1]", "error", err)
+		return 0
+	}
+
+	for i, key := range keys {
+		copy(memory[ptr+uint32(i*32):ptr+uint32((i+1)*32)], key.Encode())
+	}
+
+	ret := pointerAndSizeToInt64(int32(ptr), int32(len(keys)*32))
+	return C.int64_t(ret)
 }
 
 //export ext_crypto_ed25519_sign_version_1
-func ext_crypto_ed25519_sign_version_1(context unsafe.Pointer, a, z C.int32_t, y C.int64_t) C.int64_t {
+func ext_crypto_ed25519_sign_version_1(context unsafe.Pointer, keyTypeID C.int32_t, key C.int32_t, msg C.int64_t) C.int64_t {
 	logger.Trace("[ext_crypto_ed25519_sign_version_1] executing...")
-	logger.Warn("[ext_crypto_ed25519_sign_version_1] unimplemented")
-	return 0
+
+	instanceContext := wasm.IntoInstanceContext(context)
+	runtimeCtx := instanceContext.Data().(*runtime.Context)
+	memory := instanceContext.Memory().Data()
+
+	pubKeyData := memory[key : key+32]
+	pubKey, err := ed25519.NewPublicKey(pubKeyData)
+	if err != nil {
+		logger.Error("[ext_crypto_ed25519_sign_version_1]", "error", err)
+		return 0
+	}
+
+	signingKey := runtimeCtx.Keystore.GetKeypair(pubKey)
+	if signingKey == nil {
+		logger.Error("[ext_crypto_ed25519_sign_version_1] could not find key in keystore", "public key", pubKey)
+		return 0
+	}
+
+	sig, err := signingKey.Sign(asMemorySlice(instanceContext, msg))
+	if err != nil {
+		logger.Error("[ext_crypto_ed25519_sign_version_1] could not sign message")
+		return 0
+	}
+
+	ret, err := toWasmMemoryOptional(instanceContext, sig)
+	if err != nil {
+		logger.Error("[ext_crypto_ed25519_sign_version_1] failed to allocate memory", err)
+		return 0
+	}
+
+	return C.int64_t(ret)
 }
 
 //export ext_crypto_ed25519_verify_version_1
-func ext_crypto_ed25519_verify_version_1(context unsafe.Pointer, a C.int32_t, z C.int64_t, y C.int32_t) C.int32_t {
+func ext_crypto_ed25519_verify_version_1(context unsafe.Pointer, sig C.int32_t, msg C.int64_t, key C.int32_t) C.int32_t {
 	logger.Trace("[ext_crypto_ed25519_verify_version_1] executing...")
-	logger.Warn("[ext_crypto_ed25519_verify_version_1] unimplemented")
-	return 0
+
+	instanceContext := wasm.IntoInstanceContext(context)
+	memory := instanceContext.Memory().Data()
+	signVerify := instanceContext.Data().(*runtime.Context).SigVerifier
+
+	newSign := memory[sig : sig+64]
+	newMsg := asMemorySlice(instanceContext, msg)
+	pubKeyData := memory[key : key+32]
+
+	pubKey, err := ed25519.NewPublicKey(pubKeyData)
+	if err != nil {
+		logger.Error("[ext_crypto_ed25519_verify_version_1] failed to access public Key")
+		return 0
+	}
+
+	if signVerify.IsStarted() {
+		signature := runtime.Signature{
+			PubKey:    pubKey.Encode(),
+			Sign:      newSign,
+			Msg:       newMsg,
+			KyeTypeID: crypto.Ed25519Type,
+		}
+		signVerify.Add(&signature)
+		return 0
```

Comment:
this tells the runtime the signature is invalid, but it seems it was added to the batch? maybe I am misunderstanding this

Reply:
Add() check if it has encountered an invalid signature before. If so then it skips adding it to the batch since the result of the batch will be false anyway.